### PR TITLE
AIP-72: Handling "up_for_reschedule" task instance states

### PR DESCRIPTION
**Pull request by amoghrajesh**
related: https://github.com/apache/airflow/issues/44414

### Server side changes (execution api):
1. Handling `TIRescheduleStatePayload` in `ti_update_state` -> covered by unit test: `test_ti_update_state_to_reschedule`
2. Defining a datamodel for TIRescheduleStatePayload and adding it to the discriminator: ti_state_discriminator

#### HTTP client:
Added a new function defer that sends a patch request to the `task-instances/{id}/state execution` api with payload: `RescheduleTask`

#### Comms:
Defining a new data model to send a request to patch ti as "up_for_reschedule" from task runner to supervisor: `RescheduleTask`  (Added to ToSupervisor)

#### Supervisor:
1. Extended `handle_requests` to receive requests from task runner and forwarding the message to http client to call reschedule.

#### TaskRunner:
Task runner executes: ti.task.execute and raises `AirflowRescheduleException` for rescheduling. This sends a request to supervisor using `SUPERVISOR_COMMS`

Example DAG with which this was tested:
```
from datetime import datetime, timedelta
from airflow.exceptions import AirflowRescheduleException
from airflow.decorators import dag, task
from airflow.utils import timezone


@dag()
def reschedule_dag():
    @task
    def raise_reschedule_exception(**kwargs):
        raise AirflowRescheduleException(reschedule_date=timezone.datetime(2025, 1, 1))

    raise_reschedule_exception()


reschedule_dag()
```

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/api_fastapi/execution_api/routes/task_instances.py [hunk 1]
```diff
@@ -147,7 +149,26 @@ def ti_update_state(
             next_kwargs=ti_patch_payload.trigger_kwargs,
             trigger_timeout=timeout,
         )
+    elif isinstance(ti_patch_payload, TIRescheduleStatePayload):
+        task_instance = session.scalar(select(TI).where(TI.id == ti_id_str))
+        actual_start_date = timezone.utcnow()
+        # update the task_reschedule table first
+        session.add(
+            TaskReschedule(
```
Comment (ephraimbuddy) [inline airflow/api_fastapi/execution_api/routes/task_instances.py:237]: Are we creating a new TaskReschedule here or updating as per comment on line 155?

Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/task_instances.py:237]: We are creating a new one here. By update i meant, handle the TaskReschedule first. I will change the comment to word it better

Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/task_instances.py:237]: Should be better now @ephraimbuddy
Handled in 53c80679d878ad2f6ad9f70fb4f2304176591b1e

File: airflow/api_fastapi/execution_api/routes/task_instances.py [hunk 2]
```diff
@@ -147,7 +149,26 @@ def ti_update_state(
             next_kwargs=ti_patch_payload.trigger_kwargs,
             trigger_timeout=timeout,
         )
+    elif isinstance(ti_patch_payload, TIRescheduleStatePayload):
+        task_instance = session.scalar(select(TI).where(TI.id == ti_id_str))
```
Comment (ashb) [inline airflow/api_fastapi/execution_api/routes/task_instances.py]: Sqla has a method for getting rows my their primary key -- we should use that here.
```suggestion
        task_instance = session.get(TI, (ti_id_str,))
```

File: airflow/api_fastapi/execution_api/routes/task_instances.py [hunk 3]
```diff
@@ -147,7 +149,26 @@ def ti_update_state(
...
+                actual_start_date,
+                task_instance.end_date,
+                ti_patch_payload.reschedule_date,
+                task_instance.map_index,
+            )
+        )
 
+        query = update(TI).where(TI.id == ti_id_str)
+        # clear the next_method and next_kwargs so that none of the retries pick them up
+        query = query.values(state=State.UP_FOR_RESCHEDULE, next_method=None, next_kwargs=None)
```
Comment (ashb) [inline airflow/api_fastapi/execution_api/routes/task_instances.py:171]: We've already got the TI row loaded, so lets do this:
```suggestion
        ti.state = State.UP_FOR_RESCHEDULE
        ti.next_method = None
        ti.next_kwargs = None
        session.flush()
```

File: task_sdk/src/airflow/sdk/api/client.py [hunk 4]
```diff
@@ -134,6 +135,13 @@ def defer(self, id: uuid.UUID, msg):
         # Create a deferred state payload from msg
         self.client.patch(f"task-instances/{id}/state", content=body.model_dump_json())
 
+    def reschedule(self, id: uuid.UUID, msg):
+        """Tell the API server that this TI has been reschduled."""
+        body = TIRescheduleStatePayload(**msg.model_dump(exclude_unset=True))
```
Comment (ashb) [inline task_sdk/src/airflow/sdk/api/client.py:144]: What is msg here? It should probably have a type, or if it's just the datetime pass that in directly and create the message type in this function (similar to how we do for `heartbeat` and `finish` methods. We should likely have done `defer` differently too. Maybe.)

Comment (ashb) [reply task_sdk/src/airflow/sdk/api/client.py:144]: WDYT @kaxil ?

File: airflow/api_fastapi/execution_api/routes/task_instances.py (hunk 2 above)
Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/task_instances.py]: Oh nice. Thanks.
Made the change

File: task_sdk/src/airflow/sdk/api/client.py (hunk 4 above)
Comment (amoghrajesh) [reply task_sdk/src/airflow/sdk/api/client.py:144]: Its of `RescheduleTask` class. The reason we just passed it down to this api endpoint is so that we do not need to worry about unpacking the object in the `handle_requests`. 
```
        elif isinstance(msg, RescheduleTask):
            self._terminal_state = IntermediateTIState.UP_FOR_RESCHEDULE
            self.client.task_instances.reschedule(self.id, msg)
            resp = None
```

But a type can be added for sure. Let me do that

File: airflow/api_fastapi/execution_api/routes/task_instances.py [hunk 5]
```diff
@@ -33,9 +33,11 @@
     TIDeferredStatePayload,
     TIEnterRunningPayload,
     TIHeartbeatInfo,
+    TIRescheduleStatePayload,
     TIStateUpdate,
     TITerminalStatePayload,
 )
+from airflow.models import TaskReschedule
```
Comment (kaxil) [inline airflow/api_fastapi/execution_api/routes/task_instances.py]:
```suggestion
from airflow.models.taskreschedule import TaskReschedule
```

for consistency

File: airflow/api_fastapi/execution_api/routes/task_instances.py [hunk 6]
```diff
@@ -147,7 +149,26 @@ def ti_update_state(
             next_kwargs=ti_patch_payload.trigger_kwargs,
             trigger_timeout=timeout,
         )
+    elif isinstance(ti_patch_payload, TIRescheduleStatePayload):
+        task_instance = session.get(TI, (ti_id_str,))
```
Comment (kaxil) [inline airflow/api_fastapi/execution_api/routes/task_instances.py]:
```suggestion
        task_instance = session.get(TI, ti_id_str)
```

File: airflow/api_fastapi/execution_api/routes/task_instances.py (hunk 5 above)
Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/task_instances.py]: Sure, handled

File: airflow/api_fastapi/execution_api/routes/task_instances.py (hunk 6 above)
Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/task_instances.py]: Oh yea. Handled

File: airflow/api_fastapi/execution_api/routes/task_instances.py [hunk 7]
```diff
@@ -224,7 +226,26 @@ def ti_update_state(
...
+        actual_start_date = timezone.utcnow()
+        # add changes to TaskReschedule table to the session
+        session.add(
+            TaskReschedule(
+                task_instance.task_id,
+                task_instance.dag_id,
+                task_instance.run_id,
+                task_instance.try_number,
+                actual_start_date,
+                task_instance.end_date,
```
Comment (kaxil) [inline airflow/api_fastapi/execution_api/routes/task_instances.py]: Hmm? How would `end_date` have any value here?

Task was set to running in `run` endpoint, it was then running via Supervisor/Task Runner, now it caught `AirflowRescheduleException`, so it will now set its state to `UP_FOR_RESCHEDULE` along with all the other things on TI table  like `end_date, duration`.

You would need something similar to 
https://github.com/apache/airflow/blob/38201cf1cfa23a6a363c31f66c39529a2b9b7ae3/airflow/api_fastapi/execution_api/routes/task_instances.py#L205

Comment (kaxil) [reply airflow/api_fastapi/execution_api/routes/task_instances.py]: Similar to how it is handled today

https://github.com/apache/airflow/blob/8848d6e35f6ead2cc1ed9887af8360fbd61c78c3/airflow/models/taskinstance.py#L1621-L1630

File: task_sdk/src/airflow/sdk/api/client.py (hunk 4 above)
Comment (kaxil) [reply task_sdk/src/airflow/sdk/api/client.py:144]: Yeah, I go back and forth on it.

`RescheduleTask` is an easy one because it currently has only one attribute: `reschedule_date`, which is needed from `Task Runner` -> `Supervisor`, so it can be passed an explicit argument.

But for something like defer it has `classpath`, `trigger_kwargs`, `next_method` and `trigger_timeout`, so we just need to decide whether we want to keep client code decoupled. The downside is some code duplication

We don't strictly need to do that, as the `payload` expected by various endpoints, and therefore for client, are all in `_generated.py` which are inherited in `comms.py` anyway. So there is already a coupling of "comm messages" between Supervisor & Task Runner with the Some of the Payloads.

----

But regardless of that, yes we should have the typehints.

File: airflow/api_fastapi/execution_api/routes/task_instances.py (hunk 7 above)
Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/task_instances.py]: Thanks for the point. I agree with you that we need "end_date". I have changed the code now to handle it like it is handled for the `TITerminalStatePayload` related states.

As for adding query like `query = TI.duration_expression_update(ti_patch_payload.end_date, query, session.bind) `, we do not need that to calculate the duration because the model itself has logic to do that as long as rthe start and end dates are set properly. I have adding a unit test that can validate this part as well. https://github.com/apache/airflow/blob/main/airflow/models/taskreschedule.py#L95-L98

Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/task_instances.py]: https://github.com/apache/airflow/pull/44907/commits/80f9ddc59d60ea1f6c6e2ee260845d0b4736d5b2 handles these scenarios. I also realised that there was no UT on the task_runner side. I added that as well.

File: task_sdk/src/airflow/sdk/api/client.py (hunk 4 above)
Comment (amoghrajesh) [reply task_sdk/src/airflow/sdk/api/client.py:144]: Typehints are already added. Maybe in a follow up we can decide and perform some cleanup.

File: airflow/api_fastapi/execution_api/routes/task_instances.py (hunk 7 above)
Comment (kaxil) [reply airflow/api_fastapi/execution_api/routes/task_instances.py]: >we do not need that to calculate the duration because the model itself has logic to do that as long as rthe start and end dates are set properly. I have adding a unit test that can validate this part as well. https://github.com/apache/airflow/blob/main/airflow/models/taskreschedule.py#L95-L98

That is TaskReschedule, but we need duration for TI too

Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/task_instances.py]: Just ran an example and realised that. I will fix it.

Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/task_instances.py]: Fixed in 7cc6e9d86c1245100a56c9e78fb82269779cb827, should be good now. Added in the test to validate too

File: airflow/models/taskreschedule.py [hunk 8]
```diff
@@ -21,7 +21,17 @@
...
+    Column,
+    ForeignKeyConstraint,
+    Index,
+    Integer,
+    String,
+    asc,
+    desc,
+    select,
+    text,
+)
```
Comment (kaxil) [inline airflow/models/taskreschedule.py]: unrelated

File: tests/api_fastapi/execution_api/routes/test_task_instances.py [hunk 9]
```diff
@@ -286,6 +286,57 @@ def test_ti_update_state_to_deferred(self, client, session, create_task_instance
...
+        assert len(trs) == 1
+        assert trs[0].dag_id == "dag"
+        assert trs[0].task_id == "test_ti_update_state_to_reschedule"
+        assert trs[0].run_id == "test"
+        assert trs[0].try_number == 0
+        assert trs[0].start_date == instant
+        assert trs[0].end_date == DEFAULT_END_DATE
+        assert trs[0].reschedule_date == timezone.parse("2025-1-1T12:00:00Z")
+        assert trs[0].map_index == -1
+        assert ceil(trs[0].duration) == 129600
```
Comment (kaxil) [inline tests/api_fastapi/execution_api/routes/test_task_instances.py]: We shouldn't need `ceil` anymore

File: tests/api_fastapi/execution_api/routes/test_task_instances.py [hunk 10]
```diff
@@ -286,6 +286,57 @@ def test_ti_update_state_to_deferred(self, client, session, create_task_instance
...
+        instant = timezone.datetime(2024, 10, 30)
+        time_machine.move_to(instant, tick=False)
+
+        ti = create_task_instance(
+            task_id="test_ti_update_state_to_reschedule",
+            state=State.RUNNING,
+            session=session,
+        )
+        ti.start_date = instant
+        ti.end_date = DEFAULT_END_DATE
```
Comment (kaxil) [inline tests/api_fastapi/execution_api/routes/test_task_instances.py]: Hmm this shouldn't be set already since task is marked as running

File: tests/api_fastapi/execution_api/routes/test_task_instances.py [hunk 11]
```diff
@@ -286,6 +286,57 @@ def test_ti_update_state_to_deferred(self, client, session, create_task_instance
...
+        session.commit()
+
+        payload = {
+            "state": "up_for_reschedule",
+            "reschedule_date": timezone.parse("2025-1-1T12:00:00Z").isoformat(),
```
Comment (kaxil) [inline tests/api_fastapi/execution_api/routes/test_task_instances.py]: Why this? i.e. string to date to string?

also the start date vs reschedule date vs end date are poles apart :) 


```
DEFAULT_START_DATE = timezone.parse("2024-10-31T11:00:00Z")
DEFAULT_END_DATE = timezone.parse("2024-10-31T12:00:00Z")
reschedule_date = timezone.parse("2025-1-1T12:00:00Z")
```

File: airflow/api_fastapi/execution_api/routes/task_instances.py [hunk 12]
```diff
@@ -228,7 +230,28 @@ def ti_update_state(
             next_kwargs=ti_patch_payload.trigger_kwargs,
             trigger_timeout=timeout,
         )
+    elif isinstance(ti_patch_payload, TIRescheduleStatePayload):
+        task_instance = session.get(TI, ti_id_str)
+        actual_start_date = timezone.utcnow()
+        # add changes to TaskReschedule table to the session
```
Comment (kaxil) [inline airflow/api_fastapi/execution_api/routes/task_instances.py]:
```suggestion
```

**Review by kaxil — APPROVED** [review]
Approving (lgtm high-level), but please address the comments

File: airflow/models/taskreschedule.py (hunk 8 above)
Comment (amoghrajesh) [reply airflow/models/taskreschedule.py]: Removed

File: tests/api_fastapi/execution_api/routes/test_task_instances.py (hunk 9 above)
Comment (amoghrajesh) [reply tests/api_fastapi/execution_api/routes/test_task_instances.py]: Handled

File: tests/api_fastapi/execution_api/routes/test_task_instances.py (hunk 10 above)
Comment (amoghrajesh) [reply tests/api_fastapi/execution_api/routes/test_task_instances.py]: Good call. removing

File: tests/api_fastapi/execution_api/routes/test_task_instances.py (hunk 11 above)
Comment (amoghrajesh) [reply tests/api_fastapi/execution_api/routes/test_task_instances.py]: Yeah, haha. I was keeping the reschedule date to a much later date on purpose. Let me move it closer

> Why this? i.e. string to date to string?

No specific reason, I will just directly use the string. Direclty passing string now

File: airflow/api_fastapi/execution_api/routes/task_instances.py (hunk 12 above)
Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/task_instances.py]: Removed

File: tests/api_fastapi/execution_api/routes/test_task_instances.py (hunk 11 above)
Comment (amoghrajesh) [reply tests/api_fastapi/execution_api/routes/test_task_instances.py]: Handled in b8f81e53ff810537c0b03041f3840f9996bfafce

File: task_sdk/src/airflow/sdk/api/client.py (hunk 4 above)
Comment (amoghrajesh) [reply task_sdk/src/airflow/sdk/api/client.py:144]: Resolving for now, will revisit during cleanup

**Comment by amoghrajesh** [comment]
@kaxil merging this one. I think I have handled every comment
